compress.cc

```diff
@@ -531,3 +652,391 @@ size_t snappy_processor::compress_max_size(size_t input_len) const {
     return snappy_max_compressed_length(input_len);
 }
 
+// Constructs compressors and decompressors for SSTables,
+// making sure that the expensive identical parts (dictionaries) are shared
+// across nodes.
+//
+// Holds weak pointers to all live dictionaries
+// (so that they can be cheaply shared with new SSTables if an identical dict is requested),
+// and shared (lifetime-extending) pointers to the current writer ("recommended")
+// dict for each table (so that they can be shared with new SSTables without consulting
+// `system.dicts`).
+//
+// To make coordination work without resorting to std::mutex and such, dicts have owner shards,
+// (and are borrowed by foreign shared pointers) and all requests for a given dict ID go through its owner.
+// (Note: this shouldn't pose a performance problem because a dict is only requested once per an opening of an SSTable).
+// (Note: at the moment of this writing, one shard owns all. Later we can spread the ownership. (E.g. shard it by dict hash)).
```

Comment:
Does "Later" mean a later patch?

Reply:
(in this series)

Reply:
It's okay to do it after the series, but let's do it quickly. I can't predict how much NUMA will affect it and testing will be very hard.

Reply:
Do we even run any tests with NUMA machines?